Automated cherry pick of #105915: sched: ensure feature gate is honored when instantiating
#106105: Add unit tests to cover scheduler's setup

Huang-Wei opened the pull request:
Cherry pick of #105915 #106105 on release-1.22.

#105915: sched: ensure feature gate is honored when instantiating
#106105: Add unit tests to cover scheduler's setup

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
The --leader-elect* CLI args are now honored correctly in scheduler.
```

Huang-Wei committed:
sched: ensure feature gate is honored when instantiating scheduler

Huang-Wei committed:
Add unit tests to cover scheduler's setup

Huang-Wei commented:
/kind bug
/sig scheduling
/triage accepted
/priority important-soon

/assign @alculquicondor

alculquicondor commented:
/lgtm

Huang-Wei commented:
/hold
I will look into the test failures.

Huang-Wei commented:
/test pull-kubernetes-unit

Huang-Wei commented:
With this PR, Options#CombinedInsecureServing is always populated properly via cobra flagging - it's independent with the embedded part in defaulting CC's HealthzBindAddress. So the test failure is due to old logic overriding the Options#CombinedInsecureServing - which is no longer needed.

In terms of prioritizing insecure flags (CLI args or defaulting or values from provided --config), `CombinedInsecureServing.ApplyTo` and `CombinedInsecureServing.ApplyToFromLoadedConfig` covers it properly.

Huang-Wei commented:
/retest

alculquicondor commented:
Can you put the changes that were not automatically cherry-picked in a separate commit?

Huang-Wei commented:
> Can you put the changes that were not automatically cherry-picked in a separate commit?

It doesn't make a lot of sense doing that as even the original cherry-pick commit needs to resolve conflicts manually.

The delta changes are for insecure serving - we have to specify the default values for the CLI args (as we delayed the instantiation of the defaulting obj), also as I mentioned above, we don't need to verify if the insecure serving CLI args get changed or not - `CombinedInsecureServing.Apply...` can take care of that. Below is the diff:

```diff
diff --git a/cmd/kube-scheduler/app/options/insecure_serving.go b/cmd/kube-scheduler/app/options/insecure_serving.go
index 449ba266d91..5e3bcbc660b 100644
--- a/cmd/kube-scheduler/app/options/insecure_serving.go
+++ b/cmd/kube-scheduler/app/options/insecure_serving.go
@@ -44,10 +44,10 @@ func (o *CombinedInsecureServingOptions) AddFlags(fs *pflag.FlagSet) {
 		return
 	}
 
-	fs.StringVar(&o.BindAddress, "address", o.BindAddress, "DEPRECATED: the IP address on which to listen for the --port port (set to 0.0.0.0 or :: for listening in all interfaces and IP families). See --bind-address instead. This parameter is ignored if a config file is specified in --config.")
+	fs.StringVar(&o.BindAddress, "address", "0.0.0.0", "DEPRECATED: the IP address on which to listen for the --port port (set to 0.0.0.0 or :: for listening in all interfaces and IP families). See --bind-address instead. This parameter is ignored if a config file is specified in --config.")
 	// MarkDeprecated hides the flag from the help. We don't want that:
 	// fs.MarkDeprecated("address", "see --bind-address instead.")
-	fs.IntVar(&o.BindPort, "port", o.BindPort, "DEPRECATED: the port on which to serve HTTP insecurely without authentication and authorization. If 0, don't serve plain HTTP at all. See --secure-port instead. This parameter is ignored if a config file is specified in --config.")
+	fs.IntVar(&o.BindPort, "port", kubeschedulerconfig.DefaultInsecureSchedulerPort, "DEPRECATED: the port on which to serve HTTP insecurely without authentication and authorization. If 0, don't serve plain HTTP at all. See --secure-port instead. This parameter is ignored if a config file is specified in --config.")
 	// MarkDeprecated hides the flag from the help. We don't want that:
 	// fs.MarkDeprecated("port", "see --secure-port instead.")
 }
diff --git a/cmd/kube-scheduler/app/options/options.go b/cmd/kube-scheduler/app/options/options.go
index d199b89f356..4dad742a700 100644
--- a/cmd/kube-scheduler/app/options/options.go
+++ b/cmd/kube-scheduler/app/options/options.go
@@ -122,20 +122,6 @@ func (o *Options) ApplyDeprecated() {
 		return
 	}
 
-	hhost, hport, err := splitHostIntPort(cfg.HealthzBindAddress)
-	if err != nil {
-		return err
-	}
-	// Obtain CLI args related with insecure serving.
-	// If not specified in command line, derive the default settings from cfg.
-	insecureServing := nfs.FlagSet("insecure serving")
-	if !insecureServing.Changed("address") {
-		o.CombinedInsecureServing.BindAddress = hhost
-	}
-	if !insecureServing.Changed("port") {
-		o.CombinedInsecureServing.BindPort = hport
-	}
-
 	// Obtain deprecated CLI args. Set them to cfg if specified in command line.
 	deprecated := o.Flags.FlagSet("deprecated")
 	if deprecated.Changed("profiling") {
```

alculquicondor commented:
Thanks

/lgtm

Huang-Wei commented:
/unhold

cc @kubernetes/release-managers

puerco commented:
Thank you for backporting these @Huang-Wei 
/lgtm

puerco commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd
